### PR TITLE
Adding ability to set image architecture when creating OCI images using DockerTools

### LINKED ISSUE #200725: dockerTools buildImage and streamLayeredImage ignores architecture in config
When building multi architecture images without the need to cross compile
the architecture parameter is ignored.

The following lines explain why:
https://github.com/NixOS/nixpkgs/blob/97f6e8b3e63837ebd54e5401c1be2a3ba4d44638/pkgs/build-support/docker/default.nix#L76
https://github.com/NixOS/nixpkgs/blob/97f6e8b3e63837ebd54e5401c1be2a3ba4d44638/pkgs/build-support/docker/default.nix#L549
https://github.com/NixOS/nixpkgs/blob/97f6e8b3e63837ebd54e5401c1be2a3ba4d44638/pkgs/build-support/docker/default.nix#L868

Allow users to override the target architecture as is already done in https://github.com/NixOS/nixpkgs/blob/97f6e8b3e63837ebd54e5401c1be2a3ba4d44638/pkgs/build-support/docker/default.nix#L103
or actually allow for the override to be passed in via the config settings e.g:
```
pkgs.dockerTools.buildImage {
        name = "test";
        tag = "latest";
        config = {
          Cmd = [
            "${pkgs.bash}/bin/bash"
          ];
          Architecture = "amd64";
        };
      };
```

Edit: implemented in #201902 as follows:
```
pkgs.dockerTools.buildImage {
        name = "test";
        tag = "latest";
        config = {
          Cmd = [
            "${pkgs.bash}/bin/bash"
          ];
        };
        architecture = "amd64";
      };
```

### PR DESCRIPTION
**Pull request by ck3mp3r**
###### Description of changes
Adding ability to set image architecture when creating OCI images using DockerTools.

###### Things done
Adding optional `arch` parameter for `buildImage` and `streamLayeredImage` to allow specifying image architecture for multi-architecture builds that don't require cross-compiling.

- Built on platform(s)
  - [ ] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

This fixes https://github.com/NixOS/nixpkgs/issues/200725

### CODE REVIEW

File: pkgs/build-support/docker/default.nix [hunk 1]
```diff
@@ -513,6 +513,8 @@ rec {
       keepContentsDirlinks ? false
     , # Docker config; e.g. what command to run on the container.
       config ? null
+    , # Image architecture, defaultArch will be used if null
+      arch ? defaultArch
```
Comment (roberth) [inline pkgs/build-support/docker/default.nix:517]:
```suggestion
    , # Image architecture, defaults to the architecture of the `hostPlatform` when unset
      architecture ? defaultArch
```

Please add a test case to `nixos/tests/docker-tools.nix` and update `doc/builders/images/dockertools.section.md`.

**Review by roberth — CHANGES_REQUESTED** [review]
The idea seems good, but the implementation needs a bit of work.

File: pkgs/build-support/docker/default.nix (hunk 1 above)
Comment (ck3mp3r) [reply pkgs/build-support/docker/default.nix:517]: I just literally copied the approach that was used here: https://github.com/NixOS/nixpkgs/blob/97f6e8b3e63837ebd54e5401c1be2a3ba4d44638/pkgs/build-support/docker/default.nix#L103 hence wanting to keep the same variable names.

**Comment by ck3mp3r** [comment]
@roberth is there anything else you need me to do on this? I'm advocating the use of nixpkgs for a large project (sadly cannot elaborate more on this, Fintech related, thus requiring deterministic builds) and it kinda rests on us having these changes made... also, we are going to look into authentication during pullImage calls as it currently doesn't seem to support pulling images from private repositories.

File: pkgs/build-support/docker/default.nix (hunk 1 above)
Comment (roberth) [reply pkgs/build-support/docker/default.nix]: `defaultArch` is an internal `let` binding - a truly internal name. I don't expect such names to be particularly good ones.

`arch` on the other hand would be an exposed name, so it has to fit its domain, which is the OCI spec, which uses `architecture`. The `defaultArch` and `GOARCH` names are irrelevant.

Please change it.

Comment (ck3mp3r) [reply pkgs/build-support/docker/default.nix]: I'll change it. But by the same logic it should also be changed for `pullImage`.

Comment (ck3mp3r) [reply pkgs/build-support/docker/default.nix]: @roberth as per your request, done.

**Comment by ck3mp3r** [comment]
All I want for Christmas is for this PR to be merged 😂

**Review by roberth — APPROVED** [review]

**Comment by roberth** [comment]
@ofborg test docker-tools

**Comment by roberth** [comment]
> All I want for Christmas is for this PR to be merged joy

Well, Santa wasn't happy to hear that you didn't run your tests...

**Comment by roberth** [comment]
I've fixed an eval error, but the test still fails. Could you have a look?

**Comment by ck3mp3r** [comment]
> > All I want for Christmas is for this PR to be merged joy
> 
> Well, Santa wasn't happy to hear that you didn't run your tests after your latest commit...

Not sure what is failing as I tested all of this locally... rebase gone wrong?

**Comment by ck3mp3r** [comment]
@ofborg test docker-tools

**Comment by ck3mp3r** [comment]
@roberth I've run `nix-build . -A nixosTests.docker-tools` on a virtual machine and no failing output whatsoever. I cannot reproduce the failure ofborg is getting on x86... I've run `nix-build . -A nixosTests.login` as per documentation to ensure tests are actually running, no problem there.

According to ofborg the aarch64 run succeeds...

**Comment by roberth** [comment]
Yeah I already suspected that the aarch64-linux result was bogus. Ofborg doesn't want to get in the way when tests are missing.

Nothing's there:
```
$ nix-build --argstr system aarch64-linux -A nixosTests.docker-tools && echo ok
ok

```

Anyway, I was able to reproduce the x86 ofborg error earlier and fixed it locally. I expect ofborg to succeed any minute...

**Comment by roberth** [comment]
Not sure how many problems we fixed, but we did it! :tada:

**Comment by ck3mp3r** [comment]
Very much appreciated!!! I also feel better equipped for the next ones, thanks for your help!

**Comment by roberth** [comment]
Thanks for making the change!

**Pull request merged**